File: web/src/types/DocUnit.ts
export type DocUnit = {
  id: string
  documentnumber: string
  creationtimestamp: string

  // Original file
  fileuploadtimestamp: string
  s3path: string
  filetype: string
  filename: string
  originalFileAsHTML: string // this field is FE-only, doesn't exist in the BE data model

  // RUBRIKEN
  // - Stammdaten
  aktenzeichen: string
  gerichtstyp: string
  dokumenttyp: string
  vorgang: string
  ecli: string
  spruchkoerper: string
  entscheidungsdatum: string
  gerichtssitz: string
  rechtskraft: string
  eingangsart: string
  dokumentationsstelle: string
  region: string

  // - Kurz- & Langtexte
  entscheidungsname: string
  titelzeile: string
  leitsatz: string
  orientierungssatz: string
  tenor: string
  gruende: string
  tatbestand: string
  entscheidungsgruende: string
}

// worth adding this dependency to make this shorter? https://stackoverflow.com/a/53995901
export function buildEmptyDocUnit(): DocUnit {
  return {
    id: "",
    documentnumber: "",
    creationtimestamp: "",
    fileuploadtimestamp: "",
    s3path: "",
    filetype: "",
    filename: "",
    originalFileAsHTML: "",
    aktenzeichen: "",
    gerichtstyp: "",
    dokumenttyp: "",
    vorgang: "",
    ecli: "",
    spruchkoerper: "",
    entscheidungsdatum: "",
    gerichtssitz: "",
    rechtskraft: "",
    eingangsart: "",
    dokumentationsstelle: "",
    region: "",
    entscheidungsname: "",
    titelzeile: "",
    leitsatz: "",
    orientierungssatz: "",
    tenor: "",
    gruende: "",
    tatbestand: "",
    entscheidungsgruende: "",
  }
}

File: src/main/java/de/bund/digitalservice/ris/domain/DocUnit.java
package de.bund.digitalservice.ris.domain;

import java.time.Instant;
import java.util.Calendar;
import java.util.List;
import java.util.UUID;
import lombok.*;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.Transient;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class DocUnit {
  public static final DocUnit EMPTY = new DocUnit();

  public static DocUnit createNew(DocUnitCreationInfo docUnitCreationInfo, int documentNumber) {
    return DocUnit.builder()
        .uuid(UUID.randomUUID())
        .creationtimestamp(Instant.now())
        .documentnumber(
            docUnitCreationInfo.getDocumentationCenterAbbreviation()
                + docUnitCreationInfo.getDocumentType()
                + Calendar.getInstance().get(Calendar.YEAR)
                + String.format("%06d", documentNumber))
        .build();
  }

  @Id Long id;
  UUID uuid;
  String documentnumber;
  Instant creationtimestamp;

  // Original file
  Instant fileuploadtimestamp;
  String s3path;
  String filetype;
  String filename;

  // RUBRIKEN
  // - Stammdaten
  String aktenzeichen;
  String gerichtstyp;
  String dokumenttyp;
  String vorgang;
  String ecli;
  String spruchkoerper;
  String entscheidungsdatum;
  String gerichtssitz;
  String rechtskraft;
  String eingangsart;
  String dokumentationsstelle;
  String region;

  // - Kurz- & Langtexte
  String entscheidungsname;
  String titelzeile;
  String leitsatz;
  String orientierungssatz;
  String tenor;
  String gruende;
  String tatbestand;
  String entscheidungsgruende;

  @Transient List<PreviousDecision> previousDecisions;

  public DocUnit setPreviousDecisions(List<PreviousDecision> previousDecisions) {
    this.previousDecisions = previousDecisions;
    return this;
  }

  public boolean hasFileAttached() {
    return s3path != null;
  }
}
